Fixes LOG4NET-630 Log4Net missing Timestamp etc. in internal LogLog

mirec75 opened the pull request:
The Issue [LOG4NET-630](https://issues.apache.org/jira/browse/LOG4NET-630) is fixed with this pull request.

Now you can configure in the Web.Config / App.Config what you want to see in the internal log4net Log.
For example:

&lt;appSettings&gt;
  &lt;add key=&quot;log4net.Internal.Debug&quot; value=&quot;true&quot;/&gt;
  &lt;add key=&quot;**log4net.Internal.LogMsgPrefixPattern**&quot; value=&quot;**%date [%7processid][%3thread][%appdomain]**&quot;/&gt;
&lt;/appSettings&gt;

to get log4net internal trace messages like this:
`2019-05-22 13:51:30,266 [   5984][  3][/LM/W3SVC/1/ROOT/sw-f-ch-dn-1-132029994898417032] log4net: defaultRepositoryType [log4net.Repository.Hierarchy.Hierarchy]`
...
`2019-05-22 13:51:30,322 [   5984][  3][/LM/W3SVC/1/ROOT/sw-f-ch-dn-1-132029994898417032] log4net: Adding appender named [LogFile] to logger [root].`

dpsenner commented:
Having two rather unrelated contributions in one pull request is odd. I would prefer these to be two pull requests instead.

mirec75 commented:
Hi Dominik, with adding the %message to the pattern I was torn between adding it directly and between not, because of the fact, that the message is already preformated (log4net + level + message) and there is no Type.FullName (i.e. %logger) covertor in PatternString converting, although the type is provided in LogLog... but now you persuaded me to your proposal :)

Do you think I shall use the `LoggingEvent` and the whole Layout Patterns for formatting somehow? For a quick solution with LoggingEvent I am little bit lost in the whole complexity.  Maybe I can have a deeper look at it.
Or if I keep using the PatternString, shall I print also the %logger (type full name) between the "log4net:" + level + %logger + message without configuring it in the "LogMsgPattern"?
What do you suggest?

And according to the `ThreadIdPatternConverter` - which branch (name) shall I create for a sake of this new pull request? Shall there be a new JIRA issue for it? Thanks for your suggestions.
I added the rights to my fork for you. Feel free to update my fork with your suggestions! :)

dpsenner commented:
> Do you think I shall use the `LoggingEvent` and the whole Layout Patterns for formatting somehow?

As far as I can recall, the internal logging does not use `LoggingEvent` or other infrastructure (for a good reason).

> the message is already preformated (log4net + level + message) and there is no Type.FullName (i.e. %logger)

This is perfectly fine as far as I am concerned.

> I added the rights to my fork for you. Feel free to update my fork with your suggestions! :)

Probably not going to happen in the foreseeable future. I am happy with not spending the little spare time there is left in front of the computer but with the kids.

dpsenner commented:
Given your substantial contribution, I see it necessary that you sign an [individual contributor license agreement](https://www.apache.org/dev/new-committers-guide.html#cla). It does grant the Apache Software Foundation sufficient rights to release any work you submit under the Apache license.

mirec75 commented:
Phew! Because of the %message, etc., I had a deeper look under the log4net hood, how the `PatternLayout` cooperates with `LoggingEvent`, and how and why similar thing is not available in `PatternString`.

Well, I did it. And it is (hopefully) generic. I added the generic methods `Format(TextWriter writer, object state) `and `FormatWithState(object state)` in `PatternString` and redirected the original methods `Format(TextWriter writer)` and `Format()` to it respectively. This feature opens others the possibility to implement and use custom converters, which can handle the `object state` parameter.
Then I added 3 `LogLog` specific simple Pattern Converters converting %message (filled from `LogLog.Message`), %logger (filled from `LogLog.Source`) and %level (filled from `LogLog.Prefix`).
I updated the `LogLog` internals to cooperate with this. There is a specific factory method `public static PatternString CreateForLogLog(string pattern)` in `PatternString` simplifying creation of the `LogLog` specific `PatternString`.

I added UnitTests and tested it also in integration tests with my .Net Web Services. All I tested is working fine.

Because of this "bigger" change, I did not dare to add it to the LOG4NET-630 pull request directly and parked it in a subbranch of LOG4NET-630, named: [LogLogPatternString](https://github.com/mirec75/logging-log4net/tree/feature/LogLogPatternString). Please have a look at the branch here and let me know, what to do with it (e.g. merge to this LOG4NET-630 pull request, or whatever to do):
https://github.com/mirec75/logging-log4net/tree/feature/LogLogPatternString

> Given your substantial contribution, I see it necessary that you sign an individual contributor license agreement. It does grant the Apache Software Foundation sufficient rights to release any work you submit under the Apache license.

I'll read and sign the contributor license in the next days and let you know, but I was first eager to code the solution :)

mirec75 commented:
> Given your substantial contribution, I see it necessary that you sign an [individual contributor license agreement](https://www.apache.org/dev/new-committers-guide.html#cla). It does grant the Apache Software Foundation sufficient rights to release any work you submit under the Apache license.

Finally, I just signed and sent my ICLA to Apache.

Now I'm waiting for your answer, if I shall merge my branch "LogLogPatternString" to this pull request (branch) or what to do with my last changes - see comment above.

mirec75 commented:
I merged the "LogLogPatternString" to this pull request
